### PR TITLE
Update rust january 2024

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

Periodic rust update.

- Panic handler can no longer be marked as `extern c`.
- Fixed new rustdoc warnings about comments that were not being used.
- Replace `Err(_)` with `Err(())`.
- Remove unneeded `let _ = &fn()`.
- Ignore two new lints that I don't think we want.

### Testing Strategy

travis

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by alistair23 — APPROVED** [review]

File: tools/run_clippy.sh [hunk 1]
```diff
@@ -178,6 +178,7 @@ CLIPPY_ARGS_PEDANTIC="
 -A clippy::verbose_bit_mask
 -A clippy::large_types_passed_by_value
 -A clippy::no_mangle_with_rust_abi
+-A clippy::struct-field-names
```
Comment (jrvanwhy) [inline tools/run_clippy.sh:181]: I think this should be `clippy::struct_field_names` (the difference is `_` versus `-`).

**Review by jrvanwhy — COMMENTED** [review]
I'm not a fan of the `clippy::ignored_unit_patterns` lint changes (replacing `_` with `()`). I think it just adds visual noise to the codebase.

Definitely subjective, though.

**Comment by bradjc** [comment]
> I'm not a fan of the `clippy::ignored_unit_patterns` lint changes (replacing `_` with `()`). I think it just adds visual noise to the codebase.
> 
> Definitely subjective, though.

I somewhat agree, but I really like that it helps remove noise too. It would be nice if there was a lint for only explicit `let _ =` that do not need to exist.

**Comment by bradjc** [comment]
Updated for December.

**Review by jrvanwhy — APPROVED** [review]

**Review by hudson-ayers — DISMISSED** [review]

**Comment by bradjc** [comment]
Updated for January.

**Review by alistair23 — DISMISSED** [review]

**Review by jrvanwhy — DISMISSED** [review]

**Comment by lschuermann** [comment]
Verified that this doesn't break the `fenix` Rust Nix overlay as introduced in #3727. Usually it's expected for such external dependencies to be pinned by their hash, but we don't seem to use that here.

**Review by lschuermann — APPROVED** [review]

**Comment by bradjc** [comment]
Fixed clippy issues on new board.

**Comment by lschuermann** [comment]
> Fixed clippy issues on new board.

Oh, thanks, I was just about to push as my laptop battery died.

**Review by lschuermann — APPROVED** [review]

**Pull request merged**